src/dbnode/network/server/tchannelthrift/node/service.go

```diff
@@ -729,153 +728,277 @@ func (s *service) FetchTagged(tctx thrift.Context, req *rpc.FetchTaggedRequest)
+	if err != nil {
+		return nil, err
+	}
+	var response rpc.FetchTaggedResult_
+	response.Elements = make([]*rpc.FetchTaggedIDResult_, 0, iter.NumIDs())
+	response.Exhaustive = iter.Exhaustive()
+	for iter.HasNextIDIter() {
+		idIter, err := iter.NextIDIter(ctx)
+		if err != nil {
+			return nil, err
+		}
+		id, err := idIter.IDResult()
+		if err != nil {
+			return nil, err
+		}
+		response.Elements = append(response.Elements, id)
+		for idIter.HasNextSegments() {
+			segments, err := idIter.NextSegments(ctx)
+			if err != nil {
+				return nil, err
+			}
+			id.Segments = append(id.Segments, segments)
+		}
+	}
+
+	return &response, nil
+}
+
+func (s *service) FetchTaggedIter(ctx context.Context, req *rpc.FetchTaggedRequest) (FetchTaggedResultsIter, error) {
+	db, err := s.startReadRPCWithDB()
+	if err != nil {
+		return nil, err
+	}
+	ctx.RegisterCloser(xresource.SimpleCloserFn(func() {
+		s.readRPCCompleted()
+	}))
 
 	ns, query, opts, fetchData, err := convert.FromRPCFetchTaggedRequest(req, s.pools)
 	if err != nil {
-		s.metrics.fetchTagged.ReportError(s.nowFn().Sub(callStart))
 		return nil, tterrors.NewBadRequestError(err)
 	}
 
 	queryResult, err := db.QueryIDs(ctx, ns, query, opts)
 	if err != nil {
-		s.metrics.fetchTagged.ReportError(s.nowFn().Sub(callStart))
 		return nil, convert.ToRPCError(err)
 	}
 
-	results := queryResult.Results
-	response := &rpc.FetchTaggedResult_{
-		Exhaustive: queryResult.Exhaustive,
-		Elements:   make([]*rpc.FetchTaggedIDResult_, 0, results.Size()),
-	}
-	nsIDBytes := ns.Bytes()
+	tagEncoder := s.pools.tagEncoder.Get()
+	ctx.RegisterFinalizer(tagEncoder)
 
-	// NB(r): Step 1 if reading data then read using an asynchronous block reader,
-	// but don't serialize yet so that all block reader requests can
-	// be issued at once before waiting for their results.
-	var encodedDataResults [][][]xio.BlockReader
-	if fetchData {
-		encodedDataResults = make([][][]xio.BlockReader, results.Size())
+	return newFetchTaggedResultsIter(&fetchTaggedResultsIterOpts{
+		queryResult: &queryResult,
+		queryOpts:   &opts,
+		fetchData:   fetchData,
+		db:          db,
+		docReader:   docs.NewEncodedDocumentReader(),
+		nsID:        ns,
+		tagEncoder:  tagEncoder,
+		iOpts:       s.opts.InstrumentOptions(),
+	}), nil
+}
+
+// FetchTaggedResultsIter iterates over the results from FetchTagged
+// The iterator is not thread safe and must only be accessed from a single goroutine.
+type FetchTaggedResultsIter interface {
+
+	// NumIDs returns the total number of series IDs in the result.
+	NumIDs() int
+
+	// Exhaustive returns true if NumIDs is all IDs that the query could have returned.
+	Exhaustive() bool
+
+	// HasNextIDIter returns true if there is another series ID to process.
+	HasNextIDIter() bool
+
+	// NextIDIter returns an iterator to process the results for a series ID.
+	// HasNextIDIter must be called before each call of this method.
+	NextIDIter(ctx context.Context) (IDIter, error)
```

Comment:
nit: If you check everywhere else in the code pass our iterators usually use the naming of:
```go
Next() bool // or NextIDIter() for this current example
Current() MyNextResult // or CurrentIDIter() for this current example
```

Reply:
you can leave java, but java can't leave you